jans-cedarling/cedarling/src/authz/entities/mod.rs

```diff
@@ -9,84 +9,68 @@
 
 mod create;
 mod trait_as_expression;
+mod user;
+mod workload;
 
 #[cfg(test)]
 mod test_create;
 
-use std::collections::HashSet;
-
+use super::request::ResourceData;
 use crate::common::cedar_schema::CedarSchemaJson;
-
-use crate::authz::token_data::{AccessTokenData, IdTokenData, UserInfoTokenData};
-use crate::common::policy_store::{
-    AccessTokenEntityMetadata, ClaimMappings, PolicyStore, TokenKind, TrustedIssuer,
-};
-use crate::jwt;
+use crate::common::policy_store::{ClaimMappings, PolicyStore, TokenKind};
+use crate::jwt::Token;
 use cedar_policy::EntityUid;
-pub use create::CedarPolicyCreateTypeError;
 use create::EntityParsedTypeName;
 pub use create::CEDAR_POLICY_SEPARATOR;
 use create::{build_entity_uid, create_entity, parse_namespace_and_typename, EntityMetadata};
+use std::collections::HashSet;
 
-use super::request::ResourceData;
-use super::token_data::TokenPayload;
-
-pub(crate) type ProcessTokensResult<'a> =
-    jwt::ProcessTokensResult<'a, AccessTokenData, IdTokenData, UserInfoTokenData>;
-
-/// Create workload entity
-pub fn create_workload(
-    entity_mapping: Option<&str>,
-    policy_store: &PolicyStore,
-    data: &AccessTokenData,
-    meta: &AccessTokenEntityMetadata,
-) -> Result<cedar_policy::Entity, CedarPolicyCreateTypeError> {
-    let schema = &policy_store.schema.json;
-    let namespace = policy_store.namespace();
-    let claim_mapping = &meta.entity_metadata.claim_mapping;
+pub use create::CreateCedarEntityError;
+pub use user::*;
+pub use workload::*;
 
-    let workload_entity_meta = EntityMetadata::new(
-        EntityParsedTypeName {
-            typename: entity_mapping.unwrap_or("Workload"),
-            namespace,
-        },
-        "client_id",
-    );
+const DEFAULT_TKN_PRINCIPAL_IDENTIFIER: &str = "jti";
 
-    workload_entity_meta.create_entity(schema, data, HashSet::new(), claim_mapping)
+pub struct DecodedTokens<'a> {
+    pub access_token: Option<Token<'a>>,
+    pub id_token: Option<Token<'a>>,
+    pub userinfo_token: Option<Token<'a>>,
 }
 
 /// Create access_token entity
 pub fn create_access_token(
     entity_mapping: Option<&str>,
     policy_store: &PolicyStore,
-    data: &AccessTokenData,
-    meta: &AccessTokenEntityMetadata,
-) -> Result<cedar_policy::Entity, CedarPolicyCreateTypeError> {
+    token: &Token,
```

Comment:
Why do we use an enum here that can hold any token? It doesn't make sense.

Reply:
It's used to tag each token so we know what kind it is. Having a `token_kind` field in the TokenData struct leads to problems, specifically in this function:
```rs
// authz/entities/create.rs

/// Build [`RestrictedExpression`] based on token_claim_value.
/// It tries to find mapping and apply it to `token_claim` json value.
fn get_record_expression(
    record: &CedarSchemaRecord,
    cedar_record_type: &EntityParsedTypeName<'_>,
    token_claim: &TokenClaim,
    schema: &CedarSchemaJson,
    claim_mapping: &ClaimMappings,
) -> Result<RestrictedExpression, CreateCedarEntityError> {
    // map json value of `token_claim` to TokenPayload object (HashMap)
    let mapped_claim: TokenData =
        match claim_mapping.get_mapping(token_claim.key(), &cedar_record_type.full_type_name()) {
            Some(m) => m.apply_mapping(token_claim.value()).into(),
            // if we do not have mapping, and value is json object, return TokenPayload based on it.
            // if value is not json object, return empty value
            None => {
                if let Some(map) = token_claim.value().as_object() {
                    TokenData::from_json_map(map.to_owned()) // <----- creating a new TokenData where the kind of token doesnt matter
                } else {
                    TokenData::default()
                }
            },
        };

    let mut record_restricted_exps = Vec::new();

    for (attribute_key, entity_attribute) in record.attributes.iter() {
        let attribute_type = entity_attribute.get_type()?;

        let mapped_claim_value =
            mapped_claim
                .get_claim(attribute_key)
                .ok_or(CreateCedarEntityError::MissingClaim(
                    attribute_key.to_string(),
                ))?;

        let exp = get_expression(
            &attribute_type,
            &mapped_claim_value,
            cedar_record_type,
            schema,
            claim_mapping,
        )
        .map_err(|err| {
            CreateCedarEntityError::BuildAttribute(
                cedar_record_type.full_type_name(),
                attribute_key.to_string(),
                Box::new(err),
            )
        })?;

        record_restricted_exps.push((attribute_key.to_string(), exp));
    }

    let restricted_expression =
        RestrictedExpression::new_record(record_restricted_exps.into_iter())
            .map_err(CreateCedarEntityError::CreateRecordFromIter)?;
    Ok(restricted_expression)
}
```

Reply:
I try not to change the other functions as much as I could but the old functions are restrictive since they assume too many things on how other structs should be used instead of just creating a new struct within the module. I guess i'll just refactor this function as well.

Reply:
refactored in [18ed767](https://github.com/JanssenProject/jans/pull/10436/commits/18ed7673bc485b13e045d2159b04de12cfc8641a)